arch/x86/core/x86_mmu.c

```diff
@@ -211,951 +307,815 @@ static char get_entry_code(uint64_t value)
 
-	while (remaining) {
-		uint64_t pde = *z_x86_pd_get_pde(pd, pos);
+	for (int level = 0; level < NUM_LEVELS; level++) {
+		int index;
+		pentry_t *entryp;
 
-		if ((pde & Z_X86_MMU_P) == 0 || (pde & Z_X86_MMU_US) == 0 ||
-		    (write && (pde & Z_X86_MMU_RW) == 0)) {
-			ret = -1;
+		index = get_index(virt, level);
+		entryp = &table[index];
+
+		/* Check if we're a PTE */
+		if (level == (NUM_LEVELS - 1)) {
+			*entryp = entry_val;
 			break;
 		}
 
-		to_examine = get_table_max(pos, remaining, Z_X86_PT_AREA);
-
-		if ((pde & Z_X86_MMU_PS) == 0) {
-			/* Not a 2MB PDE. Need to check all the linked
-			 * tables for this entry
+		/* This is a non-leaf entry */
+		if ((*entryp & MMU_P) == 0U) {
+			/* Not present. Never done a mapping here yet, need
+			 * some RAM for linked tables
 			 */
-			struct x86_mmu_pt *pt;
+			void *new_table = get_page(ctx);
 
-			pt = z_x86_pde_get_pt(pde);
-			ret = x86_mmu_validate_pt(pt, pos, to_examine, write);
-			if (ret != 0) {
-				break;
+			if (new_table == NULL) {
+				return -ENOMEM;
 			}
+			*entryp = z_mem_phys_addr(new_table) | INT_FLAGS;
+			table = new_table;
 		} else {
-			ret = 0;
+			/* We fail an assertion here due to no support for
+			 * splitting existing bigpage mappings.
+			 * If the PS bit is not supported at some level (like
+			 * in a PML4 entry) it is always reserved and must be 0
+			 */
+			__ASSERT((*entryp & MMU_PS) == 0U,
+				  "large page encountered");
+			table = next_table(*entryp, level);
 		}
-
-		remaining -= to_examine;
-		pos += to_examine;
 	}
 
-	return ret;
+	return 0;
 }
 
-/* Range [addr, addr + size) must fall within the bounds of the pdpt */
-static int x86_mmu_validate_pdpt(struct x86_mmu_pdpt *pdpt, uintptr_t addr,
-				 size_t size, bool write)
-{
-	uintptr_t pos = addr;
-	size_t remaining = size;
-	int ret = 0;
-	size_t to_examine;
+/* The "hole" in the 64-bit address space for 4-level IA32e paging starts
+ * after this, up to its 1's complement
+ */
+#define MAX_IA32E_ADDR ((1ULL << 47) - 1)
 
-	while (remaining) {
-		uint64_t pdpte = *z_x86_pdpt_get_pdpte(pdpt, pos);
+/* map region virt..virt+size to phys with provided arch-neutral flags */
+int arch_mem_map(void *virt, uintptr_t phys, size_t size, uint32_t flags)
+{
+	pentry_t entry_flags = MMU_P;
+	pentry_t *ptables;
 
-		if ((pdpte & Z_X86_MMU_P) == 0) {
-			/* Non-present */
-			ret = -1;
-			break;
-		}
+	LOG_DBG("%s: %p -> %p (%zu) flags 0x%x",
+		__func__, (void *)phys, virt, size, flags);
 
 #ifdef CONFIG_X86_64
-		if ((pdpte & Z_X86_MMU_US) == 0 ||
-		    (write && (pdpte & Z_X86_MMU_RW) == 0)) {
-			ret = -1;
-			break;
-		}
-#endif
-		to_examine = get_table_max(pos, remaining, Z_X86_PD_AREA);
+	/* There's a gap in the "64-bit" address space, as 4-level paging
+	 * requires bits 48 to 63 to be copies of bit 47.
+	 */
+	__ASSERT((uintptr_t)virt + (size - 1) <= MAX_IA32E_ADDR ||
+		 (uintptr_t)virt >= ~MAX_IA32E_ADDR,
+		 "non-canonical virtual address mapping %p (size %zu)",
+		 virt, size);
```

Comment:
FWIW: there is no gap if you treat this as a signed number.  Intel's docs are confusing but if you read the original AMD books they're clear that the "high bits set" area was intended as a negative space.  So this assert is more easily expressed as e.g.:

    ((((intptr_t)virt) << 16) >> 16) == virt

Also not that the number of unused high bits changed recently on some devices with the advent of 5-level paging.

Not that you really care here just for an assert, but if this logic gets repeated elsewhere it might be worth doing the signed trick.

Reply:
I'm rebasing anyway so I'll pull that in, neat!

Reply:
Aware of the 5-level paging considerations, but we never turn that on in the CPU, it's always 4-level at the moment.

It would be very easy to add 5-level paging support to this code, but I cannot conceive of any use-cases for it in this OS so I've omitted it.